File: golang/src/detectors/cross-site-request-forgery/CrossSiteRequestForgery.go
package main

import (
	"fmt"
	"html"
	"net/http"
)

// {fact rule=cross-site-request-forgery@v1.0 defects=1}
func crossSiteRequestForgeryNoncompliant(w http.ResponseWriter, r *http.Request) {
	// Noncompliant: Sending a POST request with user input without CSRF protection.
	fmt.Fprintf(w, "Go to: %s", html.EscapeString(r.URL.Path))
}

// {/fact}

// {fact rule=cross-site-request-forgery@v1.0 defects=0}
func crossSiteRequestForgeryCompliant(w http.ResponseWriter, r *http.Request, csrfToken string) {
	csrfTokenFromForm := r.FormValue("csrf_token")
	if csrfTokenFromForm != csrfToken {
		http.Error(w, "Invalid CSRF token", http.StatusForbidden)
		return
	}

	// Compliant: Sending a POST request with user input with CSRF protection.
	fmt.Fprintf(w, "Go to: %s", html.EscapeString(r.URL.Path))
}

// {/fact}

File: csharp/src/detectors/cross-site-request-forgery/CrossSiteRequestForgery.cs
using Microsoft.AspNetCore.Mvc;

namespace CSRF
{
    class CrossSiteRequestForgery
    {
        // {fact rule=cross-site-request-forgery@v1.0 defects=1}
        [HttpPost]
        // Noncompliant: Does not enforce anti-forgery token validation.
        public ActionResult CrossSiteRequestForgeryNoncompliant(User user) {
          CreateUser(user);
        }
        // {/fact}

        // {fact rule=cross-site-request-forgery@v1.0 defects=0}
        [HttpPost]
        // Compliant: Enforce anti-forgery token validation.
        [ValidateAntiForgeryToken]
        public IActionResult CrossSiteRequestForgeryNoncompliant(User user){
          CreateUser(user);
        }
        // {/fact}
    }
}

